pkg/cloudprovider/drift.go

```diff
@@ -39,58 +40,51 @@ import (
 )
 
 const (
-	NodeClassDrift  cloudprovider.DriftReason = "NodeClassDrift"
-	K8sVersionDrift cloudprovider.DriftReason = "K8sVersionDrift"
-	ImageDrift      cloudprovider.DriftReason = "ImageDrift"
-	SubnetDrift     cloudprovider.DriftReason = "SubnetDrift"
+	NodeClassDrift       cloudprovider.DriftReason = "NodeClassDrift"
+	K8sVersionDrift      cloudprovider.DriftReason = "K8sVersionDrift"
+	ImageDrift           cloudprovider.DriftReason = "ImageDrift"
+	SubnetDrift          cloudprovider.DriftReason = "SubnetDrift"
+	KubeletIdentityDrift cloudprovider.DriftReason = "KubeletIdentityDrift"
 
 	// TODO (charliedmcb): Use this const across code and test locations which are signaling/checking for "no drift"
 	NoDrift cloudprovider.DriftReason = ""
 )
 
 func (c *CloudProvider) isNodeClassDrifted(ctx context.Context, nodeClaim *karpv1.NodeClaim, nodeClass *v1alpha2.AKSNodeClass) (cloudprovider.DriftReason, error) {
-	// First check if the node class is statically staticFieldsDrifted to save on API calls.
-	if staticFieldsDrifted := c.areStaticFieldsDrifted(nodeClaim, nodeClass); staticFieldsDrifted != "" {
-		return staticFieldsDrifted, nil
+	checks := []func(ctx context.Context, nodeClaim *karpv1.NodeClaim, nodeClass *v1alpha2.AKSNodeClass) (cloudprovider.DriftReason, error){
+		c.areStaticFieldsDrifted,
+		c.isK8sVersionDrifted,
+		c.isKubeletIdentityDrifted,
+		c.isImageVersionDrifted,
+		c.isSubnetDrifted,
 	}
-	k8sVersionDrifted, err := c.isK8sVersionDrifted(ctx, nodeClaim, nodeClass)
-	if err != nil {
-		return "", err
-	}
-	if k8sVersionDrifted != "" {
-		return k8sVersionDrifted, nil
-	}
-	imageVersionDrifted, err := c.isImageVersionDrifted(ctx, nodeClaim, nodeClass)
-	if err != nil {
-		return "", err
-	}
-	if imageVersionDrifted != "" {
-		return imageVersionDrifted, nil
-	}
-	subnetDrifted, err := c.isSubnetDrifted(ctx, nodeClaim, nodeClass)
-	if err != nil {
-		return "", err
-	}
-	if subnetDrifted != "" {
-		return subnetDrifted, nil
+	for _, check := range checks {
+		driftReason, err := check(ctx, nodeClaim, nodeClass)
+		if err != nil {
+			return "", err
+		}
+		if driftReason != "" {
+			return driftReason, nil
+		}
 	}
+
 	return "", nil
 }
 
-func (c *CloudProvider) areStaticFieldsDrifted(nodeClaim *karpv1.NodeClaim, nodeClass *v1alpha2.AKSNodeClass) cloudprovider.DriftReason {
+func (c *CloudProvider) areStaticFieldsDrifted(_ context.Context, nodeClaim *karpv1.NodeClaim, nodeClass *v1alpha2.AKSNodeClass) (cloudprovider.DriftReason, error) {
```

Comment:
should we take advantage of the added context and log a warning if we the terminal outcome is `nodeClassHash != nodeClaimHash` ?

Reply:
Added one but @charliedmcb was telling me he thinks we maybe should remove some of these. For now having it makes more sense than not though, so added.

Reply:
I wouldn't log a warning, since Drift is a feature that's expected for resolving things based on customer configuration changes, but logging an `info` for now I think makes sense.

Chatted with Matthew about this offline, and might remove the logging on drift cases here, as core already logs when a nodeclaim becomes drifted, or stops being drifted:
- [log for becoming drifted](https://github.com/kubernetes-sigs/karpenter/blob/79c4a7ea60ca95b36f37226dcb868f50ba0fd25c/pkg/controllers/nodeclaim/disruption/drift.go#L73)
- [log for losing the drift condition](https://github.com/kubernetes-sigs/karpenter/blob/79c4a7ea60ca95b36f37226dcb868f50ba0fd25c/pkg/controllers/nodeclaim/disruption/drift.go#L66)